Initial draft DPoP with Angular (#1526)
oktadev/okta-blog: _source/_posts/2024-08-09-angular-dpop-jwt.md

---
layout: blog_post
title: ""
author: alisa-duncan
by: advocate
communities: [security,javascript]
description: ""
tags: []
tweets:
- ""
- ""
- ""
image:
type: conversion
---

> We use OAuth 2.0 access tokens to request data and to perform actions within our software systems. A resource server first checks the validity of the access token before acting upon the HTTP request. What happens if the requesting party is a malicious sort who stole your token? Would the resource server honor the HTTP request? If you use a bearer token, the answer is "Yes." 

Reviewer: ```suggestion
We use OAuth 2.0 access tokens to request data and to perform actions within our software systems. A resource server first checks the validity of the access token before acting upon the HTTP request. What happens if the requesting party is a malicious sort who stole your token? Would the resource server honor the HTTP request? If you use a bearer token, the answer is "yes." 
```

{% excerpt /blog/2023/04/04/spa-auth-tokens %}

**Prerequisites**

For this tutorial, you need the following tools:
  * [Node.js](https://nodejs.org/en) v18 or greater
  * A web browser with good debugging capabilities
  * Your favorite IDE. Still searching? I like VS Code and WebStorm because they have integrated terminal windows.
  * Terminal window (if you aren't using an IDE with a built-in terminal)
  * [Git](https://git-scm.com/) and an optional [GitHub account](https://github.com/) if you want to track your changes using a source control manager
  
> **Note**
>
> This code project is best for developers familiar with web development basics.
>
> Are you following along the step-by-step code instructions in Angular? If you are an Angular newbie, start by building [your first Angular app](https://angular.io/tutorial/first-app) using the tutorial created by the Angular team.

> **JAD TODO - Let's start with an application using OpenID Connect and OAuth 2.0** We're following best-practices, that is for this Angular app (a SPA) we're using authorization code flow with PKCE. Create the Okta developer account and get app going. Sign in, following network calls and URL redirection requests. Using Org AS, and making a call to users API.

Reviewer: ```suggestion
**JAD TODO - Let's start with an application using OpenID Connect and OAuth 2.0** We're following best-practices, that is for this Angular app (a SPA) we're using authorization code flow with PKCE. Create the Okta developer account and get app going. Sign in following network calls and URL redirection requests. Use Org AS, and make a call to users API.
```

Now grab the access token and make the same call from a HTTP client. Uh-oh!

## Binding OAuth 2.0 access tokens

> If we go back to the hotel keycard example, what we want is a hotel keycard only you can use. Something that links you are the rightful user of the hotel keycard. 

Reviewer: ```suggestion
If we go back to the hotel keycard example, what we want is a hotel keycard only you can use; something that links you are the rightful user of the hotel keycard. 
```

In the OAuth world, ideally we want to link the access token to the requestor of the token in a way only the requestor can use the token. This way we know that external parties can't intercept and misuse the access token.

OAuth 2.0 defines a few methods to bind access tokens such as:
  * [Client Secret](https://oauth.net/2/client-authentication/) where the token request includes a secret value only known to the requester and the authorization server
  * [Mutual TLS Client Authentication and Certificate-Bound Access Tokens (mTLS)](https://oauth.net/2/mtls/) where the access tokens are bound to the web server's certificate
  * [Private Key JSON Web Token (JWT)](https://oauth.net/private-key-jwt/) where the requesting machine-to-machine service application uses asymmetric encryption to sign a JWT

> The methods to bind access tokens apply only to confidential clients, those that can maintain secrets. Which means the methods don't apply to public clients. Public clients are apps that run authentication code within the user's hardware, such as in Single-Page Applications (SPA) and mobile clients. Publics clients are widely used with avenues for token security exploits without careful protection. Is there an alternative?

Reviewer: ```suggestion
The methods to bind access tokens apply only to confidential clients, (those that can maintain secrets) which means the methods don't apply to public clients. Public clients are apps that run authentication code within the user's hardware, such as in Single-Page Applications (SPA) and mobile clients. Public clients are widely used with avenues for token security exploits without careful protection. Is there an alternative?
```

## Demonstrating proof of possession (DPoP)

There's now a solution for public clients and for confidential clients calling sensitive resources! By taking cues from the existing methods to bind access tokens along with enhanced browser API capabilities, OAuth 2.0 published a new specification-Demonstrating Proof of Possession. You may have heard of this idea before as the concept has been around for while. With a published spec it's now official, standardized, and supported!

> **Note**
> 
> While we'll focus on SPAs in this post, the DPoP spec applies to any OAuth client and elevates authorization security by binding the token to the requestor. Some industries, such as finance, now recommend or require DPoP to protect API calls, regardless if the calling party is a SPA, mobile, or a machine-to-machine service application.

**JAD TODO - What happens if the user tries authenticating without DPoP** 

> When using DPoP, the client uses asymmetric encryption to create the proof. The DPoP proof is a JWT that includes the URI and the HTTP method of the request along with the public key. (The proof binds the HTTP request to alleviate overpowered malicious parties who steal tokens.)???

Reviewer: ```suggestion
When using DPoP, the client uses asymmetric encryption to create the proof. The DPoP proof is a JWT that includes the URI and the HTTP method of the request along with the public key. (The proof binds the HTTP request to alleviate overpowered malicious parties who steal tokens.)
```

The client includes the proof when requesting tokens from the authorization server. This allows the authorization server to bind the access tokens to the requesting client AND the URI plus HTTP method of the call. When the client requests resources, they send the access token and the proof they are the legitimate holder of the token. **JAD TODO - CLEAN UP**

> Now that token has possession proof, the tokens are no longer bearer tokens. Only the requesting client can use the token. Clients request resources by sending the access token along with a proof demonstrating they are the legitimate holders of the access token to resource servers using a new scheme `DPoP`. Let's walk through the process.

Reviewer: ```suggestion
Now that token has possession proof, the tokens are no longer bearer tokens. Only the requesting client can use the token. Clients request resources by sending the access token along with a proof demonstrating they are the legitimate holders of the access token to resource servers using a new scheme `DPoP`. Let's walk through the process:
```
